src/backend/distributed/utils/node_metadata.c

```diff
@@ -0,0 +1,531 @@
+#include "utils/fmgroids.h"
+#include "utils/rel.h"
+#include "utils/relcache.h"
+
+
+/* default group size */
+int GroupSize = 1;
+
+
+/* local function forward declarations */
+static Datum GenerateNodeTuple(WorkerNode *workerNode);
+static WorkerNode * FindWorkerNode(char *nodeName, int32 nodePort);
+static uint32 NextGroupId(void);
+static uint32 GetMaxGroupId(void);
+static uint64 GetNodeCountInGroup(uint32 groupId);
+static List * ParseWorkerNodeFile(const char *workerNodeFilename);
+
+/* declarations for dynamic loading */
+PG_FUNCTION_INFO_V1(master_add_node);
+PG_FUNCTION_INFO_V1(master_remove_node);
+PG_FUNCTION_INFO_V1(master_initialize_node_metadata);
+PG_FUNCTION_INFO_V1(master_get_new_nodeid);
+PG_FUNCTION_INFO_V1(master_get_next_groupid);
+
+
+/*
+ * master_add_node function adds a new node to the cluster. If the node already
+ * exists, the function returns with the information about the node. If not, the
+ * following prodecure is followed while adding a node.
+ * If the groupId is not explicitly given by the user, the function picks the
+ * group that the new node should be in with respect to GroupSize. Then, the
+ * new node is inserted into the local pg_dist_node.
+ */
+Datum
+master_add_node(PG_FUNCTION_ARGS)
+{
+	text *nodeName = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	int32 groupId = PG_GETARG_INT32(2);
+	char *nodeNameString = text_to_cstring(nodeName);
+
+	Relation pgDistNode = NULL;
+	Datum nextNodeId = 0;
+	int nextNodeIdInt = 0;
+	Datum returnData = 0;
+	WorkerNode *workerNode = NULL;
+
+	/* acquire a lock so that no one can do this concurrently */
+	pgDistNode = heap_open(DistNodeRelationId(), AccessExclusiveLock);
+
+	/* check if the node already exists in the cluster */
+	workerNode = FindWorkerNode(nodeNameString, nodePort);
+	if (workerNode != NULL)
+	{
+		/* fill return data and return */
+		returnData = GenerateNodeTuple(workerNode);
+
+		/* close the heap */
+		heap_close(pgDistNode, AccessExclusiveLock);
+
+		PG_RETURN_DATUM(returnData);
+	}
+
+	/* user lets Citus to decide on the group that the newly added node should be in */
+	if (groupId == 0)
+	{
+		groupId = NextGroupId();
+	}
+	else
+	{
+		uint maxGroupId = GetMaxGroupId();
+
+		if (groupId > maxGroupId)
+		{
+			ereport(ERROR, (errmsg("you cannot add a node to a non-existing group")));
+		}
+	}
+
+	/* generate the new node id from the sequence */
+	nextNodeId = master_get_new_nodeid(NULL);
+	nextNodeIdInt = DatumGetUInt32(nextNodeId);
+
+	InsertNodeRow(nextNodeIdInt, nodeNameString, nodePort, groupId);
+
+	heap_close(pgDistNode, AccessExclusiveLock);
+
+	/* fetch the worker node, and generate the output */
+	workerNode = FindWorkerNode(nodeNameString, nodePort);
+	returnData = GenerateNodeTuple(workerNode);
+
+	PG_RETURN_CSTRING(returnData);
+}
+
+
+Datum
+master_remove_node(PG_FUNCTION_ARGS)
+{
+	text *nodeName = PG_GETARG_TEXT_P(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *nodeNameString = text_to_cstring(nodeName);
+
```

Comment:
We haven't done that in other invalidation functions, because those functions don't write to the database.